storage/repository/generic.go

```diff
@@ -0,0 +1,218 @@
+package repository
+
+import (
+	"io"
+	"strings"
+
+	"fyne.io/fyne"
+)
+
+// splitNonEmpty works exactly like strings.Split(), but only returns non-empty
+// components.
+func splitNonEmpty(str, sep string) []string {
+	components := []string{}
+	for _, v := range strings.Split(str, sep) {
+		if len(v) > 0 {
+			components = append(components, v)
+		}
+	}
+	return components
+}
+
+// GenericParent can be used as a common-case implementation of
+// HierarchicalRepository.Parent(). It will create a parent URI based on
+// IETF RFC3986.
+//
+// In short, the URI is separated into it's component parts, the path component
+// is split along instances of '/', and the trailing element is removed. The
+// result is concatenated and parsed as a new URI.
+//
+// If the URI path is empty or '/', then a duplicate of the URI is returned,
+// along with URIRootError.
+//
+// NOTE: this function should not be called except by an implementation of
+// the Repository interface - using this for unknown URIs may break.
+//
+// Since: 2.0.0
+func GenericParent(u fyne.URI) (fyne.URI, error) {
+	p := u.Path()
+
+	if p == "" || p == "/" {
+		parent, err := ParseURI(u.String())
+		if err != nil {
+			return nil, err
+		}
+		return parent, URIRootError
+	}
+
+	components := splitNonEmpty(u.Path(), "/")
+
+	newURI := u.Scheme() + "://" + u.Authority()
+
+	// there will be at least one component, since we know we don't have
+	// '/' or ''.
+	newURI += "/"
+	if len(components) > 1 {
+		newURI += strings.Join(components[:len(components)-1], "/")
+	}
+
+	// stick the query and fragment back on the end
+	q := u.Query()
+	if len(q) > 0 {
+		newURI += "?" + q
+	}
+
+	f := u.Fragment()
+	if len(f) > 0 {
+		newURI += "#" + f
+	}
+
+	// NOTE: we specifically want to use ParseURI, rather than &uri{},
+	// since the repository for the URI we just created might be a
+	// CustomURIRepository that implements it's own ParseURI.
+	return ParseURI(newURI)
+}
+
+// GenericChild can be used as a common-case implementation of
+// HierarchicalRepository.Child(). It will create a child URI by separating the
+// URI into it's component parts as described in IETF RFC 3986, then appending
+// "/" + component to the path, then concatenating the result and parsing it as
+// a new URI.
+//
+// NOTE: this function should not be called except by an implementation of
+// the Repository interface - using this for unknown URIs may break.
+//
+// Since: 2.0.0
+func GenericChild(u fyne.URI, component string) (fyne.URI, error) {
+
+	// split into components and add the new one
+	components := splitNonEmpty(u.Path(), "/")
+	components = append(components, component)
+
+	// generate the scheme, authority, and path
+	newURI := u.Scheme() + "://" + u.Authority()
+	newURI += "/" + strings.Join(components, "/")
+
+	// stick the query and fragment back on the end
+	if len(u.Query()) > 0 {
+		newURI += "?" + u.Query()
```

Comment:
nit: no need to call twice, ditto for Fragment below
```suggestion
	if q := u.Query(); len(q) > 0 {
		newURI += "?" + q
```